src/utilities/codeBlockHelpers.ts

```diff
@@ -0,0 +1,215 @@
+      }
+    }
+
+    // Advance for the next match
+    // (pos + 1) if you want overlapping matches,
+    // or (pos + query.length) if you don't want overlaps.
+    startPos = pos + 1
+  }
+
+  return allMatchedLineNumbers
+}
+
+/**
+ * For a **regex** that may span multiple lines, this returns
+ * an array of ALL line numbers (1-based) touched by each match.
+ *
+ * Example:
+ *   If a match starts on line 2 and ends on line 5, this adds
+ *   [2, 3, 4, 5] to the result array for that match.
+ */
+function getAllMatchingLineNumbersByRegExp(code: string, query: string): number[] {
+  // Build line start offsets once
+  const lineOffsets = buildLineOffsets(code)
+
+  const regExp = new RegExp(query, 'sg')
+
+  const allMatchedLineNumbers: number[] = []
+  let match
+
+  while ((match = regExp.exec(code)) !== null) {
+    // Start/end offsets of the matched text
+    const startOffset = match.index
+    const endOffset = match.index + match[0].length - 1
+
+    // Map offsets to line numbers
+    const startLine = findLineForOffset(lineOffsets, startOffset)
+    const endLine = findLineForOffset(lineOffsets, endOffset)
+
+    // Collect each line from startLine through endLine
+    for (let line = startLine; line <= endLine; line++) {
+      // Avoid duplicates
+      if (allMatchedLineNumbers[allMatchedLineNumbers.length - 1] !== line) {
+        allMatchedLineNumbers.push(line)
+      }
+    }
+  }
+
+  return allMatchedLineNumbers
+}
+
+export function getMatchingLineNumbers(code: string, query: string, isRegExpMode: boolean): number[] {
+  if (isRegExpMode) {
+    return getAllMatchingLineNumbersByRegExp(code, query)
+  } else {
+    return getAllMatchingLineNumbersByExactMatch(code, query)
+  }
+}
+
+// ================================
+// Highlighting matching characters
+// ================================
+const wrapMark = (match: string) => `<mark class="matched-term">${match}</mark>`
+
+// We only need to escape '<' and '&' characters as we can assure that the escaped output
+// will only be directly set as innerHTML and will not be concatenated with other strings
+// or used inside HTML attributes.
+const ESCAPE_REGEX = /[<&]/
+export function escapeInnerHTML(raw: string): string {
+  return raw.replace(/[&<]/g, (char: string) => char === '&' ? '&amp;' : '&lt;')
+}
+
+// Provide a fast path for escaping strings that don't contain special characters
+// especially useful for large bodies of text where the overhead of escaping is non-trivial
+// It also allows to only escape necessary characters.
+export function escapeHTMLIfNeeded(raw: string, regExp = ESCAPE_REGEX, escape = escapeInnerHTML): string {
+  return regExp.test(raw) ? escape(raw) : raw
+}
+
+export function highlightMatchingChars(code: string, query: string, isRegExpMode: boolean): string {
+  let regExp: RegExp
+  try {
+    // This RegExp mus have a capture group to work here
+    regExp = new RegExp(`((?:${isRegExpMode ? query : escapeRegExp(query)})+)`, 'sgi')
```

Comment:
question: I don't have the reference on hand, but doesn't Safari have issues with lookbehind?

Reply:
It's a non-capture group, not a look behind expression.